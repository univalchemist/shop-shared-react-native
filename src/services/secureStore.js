import { Platform } from 'react-native';
import {
  setGenericPassword,
  resetGenericPassword,
  getGenericPassword,
  setInternetCredentials,
  resetInternetCredentials,
  getInternetCredentials,
  getSupportedBiometryType,
  ACCESS_CONTROL,
} from 'react-native-keychain';

const overridedSetGenericPassword = async (username, password) => {
  const options =
    Platform.OS === 'android'
      ? {
          accessControl: ACCESS_CONTROL.USER_PRESENCE,
        }
      : null;
  return await setGenericPassword(username, password, options);
};

const saveTokens = async (username, id_token, access_token, expires_in) => {
  const password = JSON.stringify({ id_token, access_token, expires_in });
  return await overridedSetGenericPassword(username, password);
};

const fetchTokens = async () => {
  const { password, username } = await getGenericPassword();
  return password ? { ...JSON.parse(password), username } : {};
};

// Shop
const saveShopTokens = async (username, integration_token, customer_token) => {
  const { id_token, access_token, expires_in } = await fetchTokens();
  const password = JSON.stringify({
    id_token,
    access_token,
    expires_in,
    // shop tokens
    integration_token,
    customer_token,
  });

  return await overridedSetGenericPassword(username, password);
};

const clearTokens = async () => {
  return await resetGenericPassword();
};

/**
 * Stores the username and password under clientId
 * @param {string} clientId
 * @param {string} username
 * @param {string} password
 */
const saveCredentials = async (clientId, username, password) => {
  try {
    const options =
      Platform.OS === 'android'
        ? {
            accessControl: ACCESS_CONTROL.APPLICATION_PASSWORD,
          }
        : null;
    return await setInternetCredentials(clientId, username, password, options);
  } catch (e) {}
};

/**
 * @typedef {Object} Credentials
 * @property {string} username - The username
 * @property {string} password - The password
 */
/**
 * Gets the username and password stored for given clientId as key
 * @param {string} clientId
 * @return {Credentials} The credentials for the given clientId
 */
const fetchCredentials = async clientId => {
  return await getInternetCredentials(clientId);
};

const clearCredentials = async clientId => {
  return await resetInternetCredentials(clientId);
};

const checkBiometricAvailability = async () => {
  return await getSupportedBiometryType();
};

export {
  overridedSetGenericPassword,
  saveTokens,
  fetchTokens,
  clearTokens,
  saveCredentials,
  fetchCredentials,
  clearCredentials,
  saveShopTokens,
  checkBiometricAvailability,
};
