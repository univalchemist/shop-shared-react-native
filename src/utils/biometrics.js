import { useState, useEffect } from 'react';
import { Platform } from 'react-native';
import DeviceInfo from 'react-native-device-info';
import * as Keychain from 'react-native-keychain';
import { BIOMETRIC_TYPE } from '@screens/Profile/constants';

export const useBiometrics = () => {
  const [state, setState] = useState({
    biometryType: null,
  });
  useEffect(() => {
    const getBiometryType = async () => {
      const result = await Keychain.getSupportedBiometryType();
      setState({
        biometryType: result,
      });
    };

    getBiometryType();
  }, []);
  return state;
};

export const getBiometricDeviceType = () => {
  if (Platform === 'android') return 'Fingerprint';
  const touchIdIPhoneList = [
    'iphone5s',
    'iphonese',
    'iphone6',
    'iphone7',
    'iphone8',
  ];
  const deviceId = DeviceInfo.getDeviceId();
  const lowerCaseDeviceId = deviceId.toLowerCase();
  if (touchIdIPhoneList.includes(lowerCaseDeviceId))
    return BIOMETRIC_TYPE.TOUCH_ID;
  return BIOMETRIC_TYPE.FACE_ID;
};
