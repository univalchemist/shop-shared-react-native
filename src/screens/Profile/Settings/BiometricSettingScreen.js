import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { Switch, StyleSheet, Platform, Alert } from 'react-native';
import { Icon } from 'react-native-elements';
import { Box, Text } from '@wrappers/components';
import { useIntl, useTheme } from '@wrappers/core/hooks';
import { useBiometrics, Storage, getBiometricDeviceType } from '@utils';
import { fetchCredentials } from '@services/secureStore';
import { BIOMETRIC_STATUS } from '@screens/Profile/constants';

const getNoteIdList = biometricType => [
  `biometricLogin.${biometricType}.note1`,
  `biometricLogin.${biometricType}.note2`,
  `biometricLogin.${biometricType}.note3`,
];

const getStyles = theme =>
  StyleSheet.create({
    borderBottom: {
      borderBottomWidth: 0.5,
      borderBottomColor: theme.colors.gray[4],
    },
    dotIcon: {
      marginTop: 10,
      marginRight: 16,
      marginLeft: 2,
    },
  });

const BiometricSettingScreen = ({ clientId }) => {
  const intl = useIntl();
  const theme = useTheme();
  const styles = getStyles(theme);
  const [biometricStatus, setBiometricStatus] = useState(false);
  const { biometryType } = useBiometrics();
  const biometricDeviceType = getBiometricDeviceType();

  useEffect(() => {
    Storage.get(BIOMETRIC_STATUS).then(biometricStatus => {
      setBiometricStatus(JSON.parse(biometricStatus));
    });
  }, []);

  if (!biometryType && !biometricDeviceType) return null;

  const biometricTypeId = biometryType ? biometryType : biometricDeviceType;
  const noteIdList = getNoteIdList(biometricTypeId);

  const handleTriggerBiometric = async () => {
    if (!biometryType) {
      Alert.alert(
        null,
        intl.formatMessage({
          id: `biometricLogin.${biometricTypeId}.enableFail.noBiometric`,
        }),
        [
          {
            text: intl.formatMessage({ id: 'ok' }),
          },
        ],
      );
      return;
    }
    try {
      const { username, password } = await fetchCredentials(clientId);

      if (username && password) {
        Storage.save(BIOMETRIC_STATUS, 'true').then(() => {
          setBiometricStatus(true);
        });
      }
    } catch (error) {
      Alert.alert(
        intl.formatMessage({
          id: `biometricLogin.${biometricTypeId}.enableFail.title`,
        }),
        intl.formatMessage({
          id: 'pleaseTryAgain',
        }),
        [
          {
            text: intl.formatMessage({ id: 'tryAgain' }),
            onPress: handleTriggerBiometric,
          },
          {
            text: intl.formatMessage({ id: 'cancel' }),
          },
        ],
      );
    }
  };

  const handleEnablingBiometric = () => {
    if (Platform.OS === 'ios') {
      // ios
      Alert.alert(
        intl.formatMessage({
          id: `biometricLogin.${biometricTypeId}.enable.title`,
        }),
        intl.formatMessage({
          id: `biometricLogin.${biometricTypeId}.enable.message`,
        }),
        [
          {
            text: intl.formatMessage({ id: 'cancel' }),
            style: 'cancel',
          },
          {
            text: intl.formatMessage({ id: 'ok' }),
            onPress: handleTriggerBiometric,
          },
        ],
      );
    } else {
      // android
      handleTriggerBiometric();
    }
  };

  const handleDisableBiometric = () => {
    Storage.save(BIOMETRIC_STATUS, 'false').then(() => {
      setBiometricStatus(false);
    });
  };

  const handleDisablingBiometric = () => {
    Alert.alert(
      Platform.OS === 'android'
        ? null
        : intl.formatMessage({
            id: `biometricLogin.${biometricTypeId}.enable.title`,
          }),
      intl.formatMessage({
        id: `biometricLogin.${biometricTypeId}.disable.message`,
      }),
      [
        {
          text: intl.formatMessage({ id: 'cancel' }),
          style: 'cancel',
        },
        {
          text: intl.formatMessage({ id: 'disable' }),
          onPress: handleDisableBiometric,
        },
      ],
    );
  };

  const handleChange = value => {
    if (value) {
      handleEnablingBiometric();
    } else {
      handleDisablingBiometric();
    }
  };

  return (
    <Box flex={1} px={32} py={24}>
      <Box
        flexDirection="row"
        justifyContent="space-between"
        alignItems="center"
        pb={24}
        style={styles.borderBottom}
      >
        <Text>
          {intl.formatMessage({
            id: `biometricLogin.${biometricTypeId}`,
          })}
        </Text>
        <Switch value={biometricStatus} onValueChange={handleChange} />
      </Box>
      <Box py={24}>
        <Text fontSize={14}>
          {intl.formatMessage({
            id: `biometricLogin.${biometricTypeId}.desc`,
          })}
        </Text>
        {noteIdList.map(noteId => (
          <Box pt={8} flexDirection="row" key={noteId}>
            <Icon
              name="fiber-manual-record"
              size={10}
              containerStyle={styles.dotIcon}
            />
            <Text fontSize={14}>{intl.formatMessage({ id: noteId })}</Text>
          </Box>
        ))}
      </Box>
    </Box>
  );
};

const mapStateToProps = state => ({
  clientId: state.user.clientId,
});

export default connect(mapStateToProps)(BiometricSettingScreen);
