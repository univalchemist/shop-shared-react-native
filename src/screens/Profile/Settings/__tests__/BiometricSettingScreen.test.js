import React from 'react';
import BiometricSettingScreen from '../BiometricSettingScreen';
import { renderForTest } from '@testUtils';
import DeviceInfo from 'react-native-device-info';
import { Switch, Alert, Platform } from 'react-native';
import {
  act,
  fireEvent,
  flushMicrotasksQueue,
} from 'react-native-testing-library';
import * as Utils from '@utils';
import * as biometricUtils from '@utils/biometrics';
import * as secureStoreUtils from '@services/secureStore';

jest.mock('react-native/Libraries/Alert/Alert', () => ({ alert: jest.fn() }));

describe('BiometricSettingScreen', () => {
  jest.spyOn(DeviceInfo, 'getDeviceId').mockReturnValue('deviceId');
  it('should render correctly', () => {
    const component = renderForTest(<BiometricSettingScreen />);

    expect(component).toMatchSnapshot();
  });
  describe('on iOS', () => {
    Platform.OS = 'ios';
    it('should alert when disabling biometric', async () => {
      const component = renderForTest(<BiometricSettingScreen />);
      const switchButton = component.getByType(Switch);
      jest.spyOn(Utils.Storage, 'save').mockResolvedValue();

      act(() => {
        fireEvent(switchButton, 'onValueChange');
      });

      act(() => {
        Alert.alert.mock.calls[0][2][1].onPress();
      });
      await flushMicrotasksQueue();

      expect(Alert.alert).toHaveBeenCalled();
      expect(Utils.Storage.save).toHaveBeenCalled();
    });

    it('should alert when enabling biometric', async () => {
      const component = renderForTest(<BiometricSettingScreen />);
      const switchButton = component.getByType(Switch);

      act(() => {
        fireEvent(switchButton, 'valueChange', true);
      });

      act(() => {
        Alert.alert.mock.calls[1][2][1].onPress();
      });
      await flushMicrotasksQueue();

      expect(Alert.alert).toHaveBeenCalled();
    });

    it('should alert for  when enabling biometric', async () => {
      const component = renderForTest(<BiometricSettingScreen />);
      const switchButton = component.getByType(Switch);

      act(() => {
        fireEvent(switchButton, 'valueChange', true);
      });

      act(() => {
        Alert.alert.mock.calls[1][2][1].onPress();
      });
      await flushMicrotasksQueue();

      expect(Alert.alert).toHaveBeenCalled();
    });
  });

  describe('on Android', () => {
    beforeEach(() => {
      Platform.OS = 'android';
    });

    it('should alert fail when enabling biometric without biometric', async () => {
      const component = renderForTest(<BiometricSettingScreen />);
      const switchButton = component.getByType(Switch);
      jest
        .spyOn(biometricUtils, 'useBiometrics')
        .mockReturnValue({ biometryType: 'FaceID' });
      jest
        .spyOn(biometricUtils, 'getBiometricDeviceType')
        .mockReturnValue('Fingerprint');

      act(() => {
        fireEvent(switchButton, 'valueChange', true);
      });
      expect(Alert.alert).toHaveBeenCalled();
    });

    it('should fetchCredentials when enabling biometric', async () => {
      const component = renderForTest(<BiometricSettingScreen />);
      const switchButton = component.getByType(Switch);
      jest.spyOn(Utils.Storage, 'save').mockResolvedValue();
      jest
        .spyOn(biometricUtils, 'useBiometrics')
        .mockReturnValue({ biometryType: 'Fingerprint' });
      jest
        .spyOn(biometricUtils, 'getBiometricDeviceType')
        .mockReturnValue('Fingerprint');
      jest
        .spyOn(secureStoreUtils, 'fetchCredentials')
        .mockResolvedValue({ username: 'username', password: 'P@ssw0rd' });

      act(() => {
        fireEvent(switchButton, 'valueChange', true);
      });

      act(() => {
        Alert.alert.mock.calls[1][2][1].onPress();
      });

      await flushMicrotasksQueue();

      expect(Utils.Storage.save).toHaveBeenCalled();
      expect(secureStoreUtils.fetchCredentials).toHaveBeenCalled();
    });
  });
});
