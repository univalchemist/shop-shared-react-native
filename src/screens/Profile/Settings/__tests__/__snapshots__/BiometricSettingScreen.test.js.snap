// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BiometricSettingScreen should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              enabled={false}
              hitSlop={
                Object {
                  "right": -700,
                }
              }
              maxDeltaY={20}
              minOffsetX={5}
              onHandlerStateChange={[Function]}
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        flex={1}
                        px={32}
                        py={24}
                        style={
                          Array [
                            Object {
                              "flexBasis": 0,
                              "flexGrow": 1,
                              "flexShrink": 1,
                              "paddingBottom": 24,
                              "paddingLeft": 32,
                              "paddingRight": 32,
                              "paddingTop": 24,
                            },
                          ]
                        }
                      >
                        <View
                          alignItems="center"
                          flexDirection="row"
                          justifyContent="space-between"
                          pb={24}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "justifyContent": "space-between",
                                "paddingBottom": 24,
                              },
                              Object {
                                "borderBottomColor": "#b8b8b8",
                                "borderBottomWidth": 0.5,
                              },
                            ]
                          }
                        >
                          <Text
                            color="gray.1"
                            fontFamily="default"
                            fontSize={16}
                            lineHeight={22}
                            style={
                              Array [
                                Object {
                                  "color": "#646464",
                                  "fontFamily": "Arial",
                                  "fontSize": 16,
                                  "lineHeight": 22,
                                },
                              ]
                            }
                          >
                            Face ID
                          </Text>
                          <RCTSwitch
                            accessibilityRole="button"
                            onChange={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "height": 31,
                                "width": 51,
                              }
                            }
                            value={false}
                          />
                        </View>
                        <View
                          py={24}
                          style={
                            Array [
                              Object {
                                "paddingBottom": 24,
                                "paddingTop": 24,
                              },
                            ]
                          }
                        >
                          <Text
                            color="gray.1"
                            fontFamily="default"
                            fontSize={14}
                            lineHeight={22}
                            style={
                              Array [
                                Object {
                                  "color": "#646464",
                                  "fontFamily": "Arial",
                                  "fontSize": 14,
                                  "lineHeight": 22,
                                },
                              ]
                            }
                          >
                            Please note the following information when you enable Face ID:
                          </Text>
                          <View
                            flexDirection="row"
                            pt={8}
                            style={
                              Array [
                                Object {
                                  "flexDirection": "row",
                                  "paddingTop": 8,
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginLeft": 2,
                                  "marginRight": 16,
                                  "marginTop": 10,
                                }
                              }
                            >
                              <View
                                replaceTheme={[Function]}
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "transparent",
                                    "justifyContent": "center",
                                  }
                                }
                                theme={
                                  Object {
                                    "colors": Object {
                                      "disabled": "hsl(208, 8%, 90%)",
                                      "divider": "#bcbbc1",
                                      "error": "#ff190c",
                                      "grey0": "#393e42",
                                      "grey1": "#43484d",
                                      "grey2": "#5e6977",
                                      "grey3": "#86939e",
                                      "grey4": "#bdc6cf",
                                      "grey5": "#e1e8ee",
                                      "greyOutline": "#bbb",
                                      "platform": Object {
                                        "android": Object {
                                          "error": "#f44336",
                                          "primary": "#2196f3",
                                          "secondary": "#9C27B0",
                                          "success": "#4caf50",
                                          "warning": "#ffeb3b",
                                        },
                                        "ios": Object {
                                          "error": "#ff3b30",
                                          "primary": "#007aff",
                                          "secondary": "#5856d6",
                                          "success": "#4cd964",
                                          "warning": "#ffcc00",
                                        },
                                      },
                                      "primary": "#2089dc",
                                      "searchBg": "#303337",
                                      "secondary": "#8F0CE8",
                                      "success": "#52c41a",
                                      "warning": "#faad14",
                                    },
                                  }
                                }
                                underlayColor="white"
                                updateTheme={[Function]}
                              >
                                <Text
                                  allowFontScaling={false}
                                  style={
                                    Array [
                                      Object {
                                        "color": "black",
                                        "fontSize": 10,
                                      },
                                      Object {
                                        "backgroundColor": "transparent",
                                      },
                                      Object {
                                        "fontFamily": "Material Icons",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      Object {},
                                    ]
                                  }
                                  testID="iconIcon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <Text
                              color="gray.1"
                              fontFamily="default"
                              fontSize={14}
                              lineHeight={22}
                              style={
                                Array [
                                  Object {
                                    "color": "#646464",
                                    "fontFamily": "Arial",
                                    "fontSize": 14,
                                    "lineHeight": 22,
                                  },
                                ]
                              }
                            >
                              Any face registered with Face ID on this device will have access to this app.
                            </Text>
                          </View>
                          <View
                            flexDirection="row"
                            pt={8}
                            style={
                              Array [
                                Object {
                                  "flexDirection": "row",
                                  "paddingTop": 8,
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginLeft": 2,
                                  "marginRight": 16,
                                  "marginTop": 10,
                                }
                              }
                            >
                              <View
                                replaceTheme={[Function]}
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "transparent",
                                    "justifyContent": "center",
                                  }
                                }
                                theme={
                                  Object {
                                    "colors": Object {
                                      "disabled": "hsl(208, 8%, 90%)",
                                      "divider": "#bcbbc1",
                                      "error": "#ff190c",
                                      "grey0": "#393e42",
                                      "grey1": "#43484d",
                                      "grey2": "#5e6977",
                                      "grey3": "#86939e",
                                      "grey4": "#bdc6cf",
                                      "grey5": "#e1e8ee",
                                      "greyOutline": "#bbb",
                                      "platform": Object {
                                        "android": Object {
                                          "error": "#f44336",
                                          "primary": "#2196f3",
                                          "secondary": "#9C27B0",
                                          "success": "#4caf50",
                                          "warning": "#ffeb3b",
                                        },
                                        "ios": Object {
                                          "error": "#ff3b30",
                                          "primary": "#007aff",
                                          "secondary": "#5856d6",
                                          "success": "#4cd964",
                                          "warning": "#ffcc00",
                                        },
                                      },
                                      "primary": "#2089dc",
                                      "searchBg": "#303337",
                                      "secondary": "#8F0CE8",
                                      "success": "#52c41a",
                                      "warning": "#faad14",
                                    },
                                  }
                                }
                                underlayColor="white"
                                updateTheme={[Function]}
                              >
                                <Text
                                  allowFontScaling={false}
                                  style={
                                    Array [
                                      Object {
                                        "color": "black",
                                        "fontSize": 10,
                                      },
                                      Object {
                                        "backgroundColor": "transparent",
                                      },
                                      Object {
                                        "fontFamily": "Material Icons",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      Object {},
                                    ]
                                  }
                                  testID="iconIcon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <Text
                              color="gray.1"
                              fontFamily="default"
                              fontSize={14}
                              lineHeight={22}
                              style={
                                Array [
                                  Object {
                                    "color": "#646464",
                                    "fontFamily": "Arial",
                                    "fontSize": 14,
                                    "lineHeight": 22,
                                  },
                                ]
                              }
                            >
                              If you tap cancel at the Face ID prompt, you can still login using your username and password
                            </Text>
                          </View>
                          <View
                            flexDirection="row"
                            pt={8}
                            style={
                              Array [
                                Object {
                                  "flexDirection": "row",
                                  "paddingTop": 8,
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginLeft": 2,
                                  "marginRight": 16,
                                  "marginTop": 10,
                                }
                              }
                            >
                              <View
                                replaceTheme={[Function]}
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "transparent",
                                    "justifyContent": "center",
                                  }
                                }
                                theme={
                                  Object {
                                    "colors": Object {
                                      "disabled": "hsl(208, 8%, 90%)",
                                      "divider": "#bcbbc1",
                                      "error": "#ff190c",
                                      "grey0": "#393e42",
                                      "grey1": "#43484d",
                                      "grey2": "#5e6977",
                                      "grey3": "#86939e",
                                      "grey4": "#bdc6cf",
                                      "grey5": "#e1e8ee",
                                      "greyOutline": "#bbb",
                                      "platform": Object {
                                        "android": Object {
                                          "error": "#f44336",
                                          "primary": "#2196f3",
                                          "secondary": "#9C27B0",
                                          "success": "#4caf50",
                                          "warning": "#ffeb3b",
                                        },
                                        "ios": Object {
                                          "error": "#ff3b30",
                                          "primary": "#007aff",
                                          "secondary": "#5856d6",
                                          "success": "#4cd964",
                                          "warning": "#ffcc00",
                                        },
                                      },
                                      "primary": "#2089dc",
                                      "searchBg": "#303337",
                                      "secondary": "#8F0CE8",
                                      "success": "#52c41a",
                                      "warning": "#faad14",
                                    },
                                  }
                                }
                                underlayColor="white"
                                updateTheme={[Function]}
                              >
                                <Text
                                  allowFontScaling={false}
                                  style={
                                    Array [
                                      Object {
                                        "color": "black",
                                        "fontSize": 10,
                                      },
                                      Object {
                                        "backgroundColor": "transparent",
                                      },
                                      Object {
                                        "fontFamily": "Material Icons",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      Object {},
                                    ]
                                  }
                                  testID="iconIcon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <Text
                              color="gray.1"
                              fontFamily="default"
                              fontSize={14}
                              lineHeight={22}
                              style={
                                Array [
                                  Object {
                                    "color": "#646464",
                                    "fontFamily": "Arial",
                                    "fontSize": 14,
                                    "lineHeight": 22,
                                  },
                                ]
                              }
                            >
                              By enabling Face ID you agree to these Terms and Condition
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
